Bounds on `<F as AsyncFnOnce>::CallOnceFuture` make the compiler assume `CallOnceFuture = ()` (?)

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![allow(warnings)]
#![feature(async_closure, async_fn_traits, unboxed_closures)]
use std::ops::AsyncFnOnce;
use std::future::Future;

trait Anything {}
impl<T: ?Sized> Anything for T {}

fn bar<F>(f: F)
where
    F: AsyncFnOnce() + 'static,
    <F as AsyncFnOnce<()>>::CallOnceFuture: Anything, // comment out this bound for no error
{
    // body not relevant, but this proves that `F::CallOnceFuture` does implement `Future`.
    let _ = async { f().await; };
}

fn main() {
    // error only occurs if `bar` is called
    bar(async move || {});
}
```

[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=725c0bfb9598ea4f7688a3ba11bd4fcc)

I expected to see this happen: This should compile successfully; `AsyncFnOnce::CallOnceFuture` is guaranteed to implement `Anything` (since there is a blanket impl).

Instead, this happened: The bound appears to cause the compiler to think that `F::CallOnceFuture = ()`, see error message:

```rs
error[E0271]: type mismatch resolving `<{async closure@main.rs:19:9} as AsyncFnOnce<()>>::CallOnceFuture == ()`
  --> src/main.rs:19:5
   |
19 |     bar(async move || {});
   |     ^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `async` closure body
   |
   = note:         expected unit type `()`
           found `async` closure body `{async closure body@src/main.rs:19:23: 19:25}`

For more information about this error, try `rustc --explain E0271`.
error: could not compile `asyncfnonce-thing` (bin "asyncfnonce-thing") due to 1 previous error

```

It compiles with no error if:
* `bar` is never called
* or we have an empty bound `<F as AsyncFnOnce<()>>::CallOnceFuture: ` instead of `: Anything`

It doesn't compile if we have *any* non-empty bound, e.g. `<F as AsyncFnOnce<()>>::CallOnceFuture: Future`. (The original issue was someone on discord trying to have a `: 'static` bound, but I used `: Anything` in the issue here since it should always be satisfied, and any bound appears to cause the issue).

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.85.0-nightly (8dc83770f 2024-12-06)
binary: rustc
commit-hash: 8dc83770f748c6cd16b342889ca2240397c19534
commit-date: 2024-12-06
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.5
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
(no ICE, so no backtrace)

@rustbot label +F-async_closure +requires-nightly

--- a/compiler/rustc_hir_typeck/src/closure.rs
+++ b/compiler/rustc_hir_typeck/src/closure.rs
@@ -454,20 +454,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         closure_kind: hir::ClosureKind,
         projection: ty::PolyProjectionPredicate<'tcx>,
     ) -> Option<ExpectedSig<'tcx>> {
-        let tcx = self.tcx;
-
-        let trait_def_id = projection.trait_def_id(tcx);
+        let def_id = projection.projection_def_id();
 
         // For now, we only do signature deduction based off of the `Fn` and `AsyncFn` traits,
         // for closures and async closures, respectively.
         match closure_kind {
-            hir::ClosureKind::Closure
-                if self.tcx.fn_trait_kind_from_def_id(trait_def_id).is_some() =>
-            {
+            hir::ClosureKind::Closure if self.tcx.is_lang_item(def_id, LangItem::FnOnceOutput) => {
                 self.extract_sig_from_projection(cause_span, projection)
             }
             hir::ClosureKind::CoroutineClosure(hir::CoroutineDesugaring::Async)
-                if self.tcx.async_fn_trait_kind_from_def_id(trait_def_id).is_some() =>
+                if self.tcx.is_lang_item(def_id, LangItem::AsyncFnOnceOutput) =>
             {
                 self.extract_sig_from_projection(cause_span, projection)
             }
@@ -475,7 +471,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // `F: FnOnce() -> Fut, Fut: Future<Output = T>` style bound. Let's still
             // guide inference here, since it's beneficial for the user.
             hir::ClosureKind::CoroutineClosure(hir::CoroutineDesugaring::Async)
-                if self.tcx.fn_trait_kind_from_def_id(trait_def_id).is_some() =>
+                if self.tcx.is_lang_item(def_id, LangItem::FnOnceOutput) =>
             {
                 self.extract_sig_from_projection_and_future_bound(cause_span, projection)
             }

--- a/tests/ui/async-await/async-closures/call-once-deduction.rs
+++ b/tests/ui/async-await/async-closures/call-once-deduction.rs
@@ -1,0 +1,14 @@
+//@ edition: 2021
+//@ check-pass
+
+#![feature(async_closure, async_fn_traits, unboxed_closures)]
+
+fn bar<F, O>(_: F)
+where
+    F: AsyncFnOnce<(), CallOnceFuture = O>,
+{
+}
+
+fn main() {
+    bar(async move || {});
+}

